x-pack/plugins/cases/server/connectors/cases/cases_connector.ts

```diff
@@ -418,6 +458,115 @@ export class CasesConnector extends SubActionConnector<
+      ? await this.reopenClosedCases(casesClient, entriesWithClosedCases, casesMap)
+      : await this.createNewCasesOutOfClosedCases(
+          params,
+          casesClient,
+          entriesWithClosedCases,
+          casesMap
+        );
+
+    /**
+     * The initial map contained the closed cases. We need to remove them to
+     * avoid attaching alerts to a close case
+     */
+    return new Map([...res].filter(([_, record]) => record.theCase.status !== CaseStatuses.closed));
+  }
+
+  private async reopenClosedCases(
+    casesClient: CasesClient,
+    closedCasesEntries: GroupedAlertsWithCases[],
+    casesMap: Map<string, GroupedAlertsWithCases>
+  ): Promise<Map<string, GroupedAlertsWithCases>> {
+    const casesMapWithClosedCasesOpened = new Map(casesMap);
+
+    const bulkUpdateReq = closedCasesEntries.map((entry) => ({
+      id: entry.theCase.id,
+      version: entry.theCase.version,
+      status: CaseStatuses.open,
+    }));
+
+    /**
+     * TODO: bulkUpdate throws an error. Retry on errors.
+     */
+    const bulkUpdateCasesResponse = await casesClient.cases.bulkUpdate({ cases: bulkUpdateReq });
+
+    for (const res of bulkUpdateCasesResponse) {
+      if (casesMap.has(res.id)) {
+        const data = casesMap.get(res.id) as GroupedAlertsWithCases;
+        casesMapWithClosedCasesOpened.set(res.id, { ...data, theCase: res });
+      }
+    }
+
+    return casesMapWithClosedCasesOpened;
+  }
+
+  private async createNewCasesOutOfClosedCases(
+    params: CasesConnectorRunParams,
+    casesClient: CasesClient,
+    closedCasesEntries: GroupedAlertsWithCases[],
+    casesMap: Map<string, GroupedAlertsWithCases>
+  ): Promise<Map<string, GroupedAlertsWithCases>> {
+    const casesMapWithClosedCasesOpened = new Map(casesMap);
+    const casesMapAsArray = Array.from(casesMap.values());
+
+    const findEntryByOracleRecord = (oracleId: string) => {
+      return casesMapAsArray.find((record) => record.oracleRecord.id === oracleId);
+    };
+
+    const bulkUpdateOracleValidRecords = await this.increaseOracleRecordCounter(
+      closedCasesEntries.map((entry) => entry.oracleRecord)
+    );
+
+    const groupedAlertsWithOracleRecords = new Map<string, GroupedAlertsWithOracleRecords>();
+
+    for (const record of bulkUpdateOracleValidRecords) {
+      const foundRecord = findEntryByOracleRecord(record.id);
+
+      if (foundRecord) {
+        groupedAlertsWithOracleRecords.set(record.id, {
+          oracleKey: record.id,
+          oracleRecord: foundRecord.oracleRecord,
+          alerts: foundRecord.alerts,
+          grouping: foundRecord.grouping,
+        });
+      }
+    }
+
+    const groupedAlertsWithCaseId = this.generateCaseIds(params, groupedAlertsWithOracleRecords);
+    const bulkCreateReq = Array.from(groupedAlertsWithCaseId.values()).map((record) =>
+      this.getCreateCaseRequest(params, record)
+    );
+
+    /**
+     * TODO: bulkCreate throws an error. Retry on errors.
```

Comment:
I will handle all errors and retries in this PR https://github.com/elastic/kibana/pull/173012. I think it is better to do it on another PR as the code can grow a lot and be difficult to follow. In general, we want each action to be idempotent. This means that if we retry it will not affect the correctness of the case action. In your example, if the case exists this means that another Kibana node (that runs the case action) created the case before this Kibana node managed to do it. If we retry, on the next round the node will find that the case exists and will attach the alerts to that case without trying to create a new one. Retries are useful to break race conditions or transient errors.

Reply:
> I will handle all errors and retries in this PR #173012. I think it is better to do it on another PR as the code can grow a lot and be difficult to follow.

I'm ok with this.

> In general, we want each action to be idempotent. This means that if we retry it will not affect the correctness of the case action. In your example, if the case exists this means that another Kibana node (that runs the case action) created the case before this Kibana node managed to do it. If we retry, on the next round the node will find that the case exists and will attach the alerts to that case without trying to create a new one. Retries are useful to break race conditions or transient errors.

Won't it be idempotent anyway? no matter the path followed we always end up with these alerts attached to the case with that ID. Be it if we created the case or if other node did.

Reply:
You are right on this. It may save us a retry round but I think it will make the code more difficult to follow and the scenario of a conflict should be rare. What do you think of leaving it as it is and when we have the whole picture (retry logic, error handling, etc.) and test a lot then we can see if it is worth the optimization?

Reply:
I am ok with this, my initial suggestion was more about the scope rather than the optimization.

I just elaborated further to make sure I didn't miss some logic in your response 👍